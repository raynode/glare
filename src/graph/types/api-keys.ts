import { GQLBuild, GQLSchemaBuilder } from 'graph/builder'

import { ApiKeys, User } from 'db/models'
import { createService } from 'graph/base-service'

export const apiKeyBuild = (build: GQLBuild) => {
  build.addType('ApiKey', {
    fields: {
      key: 'String!',
      service: 'String!',
      User: 'User!',
    },
  })

  build.extendType<User>('User', {
    fields: {
      apiKeys: {
        args: {
          service: 'String',
        },
        type: '[ApiKey!]!',
      },
    },
    resolver: {
      apiKeys: async (user, args, context) =>
        ApiKeys.find(context, {
          where: query => query.where({ ...args, userId: user.id }),
        }),
    },
  })
}
