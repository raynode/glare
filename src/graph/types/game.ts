import { Build } from 'gram'

import { Games } from 'db/models'
import { createService } from 'graph/base-service'

export default <BuildMode, Context>(build: Build<BuildMode, Context>) => {
  build.addType('Game', {
    type: 'interface',
    fields: {
      name: 'String!',
      levels: '[GameLevel!]!',
      worlds: '[GameWorld!]!',
    },
  })

  build.addType('GameWorld', {
    type: 'interface',
    fields: {
      name: 'String!',
      game: 'Game!',
      levels: '[GameLevel!]!',
    },
  })
  build.addType('GameLevel', {
    type: 'interface',
    fields: {
      name: 'String',
      game: 'Game!',
      world: 'GameWorld',
      data: 'JSON!',
    },
  })

  build.addQuery(
    'getGame',
    {
      type: 'Game',
      args: {
        name: 'String!',
      },
    },
    async (_, { name }) => {
      const games = await Games.find({
        where: query => query.where({ name }),
      })
      const result = {
        ...games[0],
        __typename: name,
        resolveType: name,
      }
      console.log(name, result)
      return result
    },
  )
}
