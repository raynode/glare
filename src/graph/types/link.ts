import { GQLBuild, GQLSchemaBuilder } from 'graph/builder'
import { GraphQLBoolean, GraphQLEnumType, GraphQLList, GraphQLString } from 'graphql'

import { Links } from 'db/models'
import { createService } from 'graph/base-service'

export const linkBuilder = (builder: GQLSchemaBuilder) => {
  const link = builder.model('Link', createService(Links))
  link.attr('datetime', builder.getScalar('DateTime'))
  link.attr('title', GraphQLString)
  link.attr('url', GraphQLString)
  link.attr('content', GraphQLString)
  link.attr('tags', GraphQLList(GraphQLString))
}

export default linkBuilder
