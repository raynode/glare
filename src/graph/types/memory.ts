import { Build } from 'gram'
import { getMemoryHumanReadable } from 'services/memory'

export const memoryBuild = <BuildMode, Context>(build: Build<BuildMode, Context>) => {
  build.addQuery('getMemory', 'Memory!', { resolver: getMemoryHumanReadable })
  build.addType('Memory', {
    fields: {
      rss: 'String!',
      heapTotal: 'String!',
      heapUsed: 'String!',
      external: 'String!',
    },
  })
}
