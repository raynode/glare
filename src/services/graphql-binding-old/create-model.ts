import { capitalize, pluralize, singularize } from 'inflection'
import { reduce } from 'lodash'
import { Associations, Attribute, Attributes, BaseModel, Methods, SimpleAttribute, SimpleAttributes } from './types'

export const createAttribute = (attribute: SimpleAttribute): Attribute => ({
  allowNull: true,
  allowUpdate: true,
  comment: '',
  defaultValue: undefined,
  type: 'scalar',
  visible: true,
  ...attribute,
})

export const convertSimpleAttributes = <Type>(simpleAttributes: SimpleAttributes<Type>) =>
  reduce(
    simpleAttributes,
    (memo, simpleAttribute, key) => {
      memo[key] = createAttribute(simpleAttribute)
      return memo
      // tslint:disable-next-line no-object-literal-type-assertion
    },
    {} as Attributes<Type>,
  )

export const createModel = <Type = any>(
  name: string,
  attributes: SimpleAttributes<Type>,
  associations?: Associations<Type>,
): BaseModel<Type> => {
  return {
    associations,
    assocResolvers: [],
    attributes: convertSimpleAttributes(attributes),
    fieldNames: {
      create: `${name}CreateInput`,
      list: `${name}ListType`,
      model: `${name}`,
      order: `${name}OrderEnum`,
      page: `Page`,
      update: `${name}UpdateInput`,
    },
    inspect: () => `${name}Model`,
    name,
    names: {
      create: `onCreate${name}`,
      createOne: `create${name}`,
      delete: `onDelete${name}`,
      deleteMany: `delete${pluralize(name)}`,
      findMany: `${pluralize(name)}`,
      findOne: `${name}`,
      filters: `${name}Filter`,
      updateData: `Update${name}`,
      createData: `Initial${name}`,
      update: `onUpdate${name}`,
      updateOne: `update${name}`,
    },
  }
}
