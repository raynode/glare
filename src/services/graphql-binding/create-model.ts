import { capitalize, pluralize, singularize } from 'inflection'

import { Attributes, BaseModel, Methods } from './types'

export const createModel = <Type = any>(name: string, attributes: Attributes): BaseModel<Type> => {
  return {
    associations: {},
    assocResolvers: [],
    attributes,
    fieldNames: {
      create: `${name}CreateInput`,
      list: `${name}ListType`,
      model: `${name}`,
      order: `${name}OrderEnum`,
      page: `Page`,
      update: `${name}UpdateInput`,
    },
    inspect: () => `${name}Model`,
    name,
    names: {
      create: `onCreate${name}`,
      createOne: `create${name}`,
      delete: `onDelete${name}`,
      deleteMany: `delete${pluralize(name)}`,
      findMany: `${pluralize(name)}`,
      findOne: `${name}`,
      filters: `${name}Filter`,
      updateData: `Update${name}`,
      createData: `Initial${name}`,
      update: `onUpdate${name}`,
      updateOne: `update${name}`,
    },
  }
}
