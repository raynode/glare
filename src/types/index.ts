import { create } from 'services/logger'
import { account } from 'types/account'
import { BaseBlock, Block, TypeDef } from 'types/def'
import { expense } from 'types/expense'
import { post } from 'types/post'
import { root } from 'types/root'
import { tag } from 'types/tag'
import { user } from 'types/user'

// import { widget } from 'types/widget'
// import { widgetGroup } from 'types/widget-group'

const log = create('types')

const all = [
  root,
  // account,
  // expense,
  user,
  // tag,
  // # widget,
  // post,
  // # widgetGroup,
]

const extendBaseBlock = (base: BaseBlock<any>) => (name: string) => (block: Block<any>) => {
  log(`extending ${block.name} with ${name}`)
  base.Query = { ...base.Query, ...block.Query }
  base.Mutation = { ...base.Mutation, ...block.Mutation }
  base.Resolver = {
    ...base.Resolver,
    [block.name]: { ...base.Resolver[block.name], ...block.Resolver },
  }
  base.Subscription = { ...base.Subscription, ...block.Subscription }
  return base
}

export const resolvers = all.reduce(
  (memo: BaseBlock<any>, block) => {
    const extender = extendBaseBlock(memo)
    memo.typeDefs.push(block.typeDefs)
    if (block.joins) block.joins.forEach(extender(block.name))
    return extender('root')(block)
  },
  {
    name: 'root',
    Query: {},
    Resolver: {},
    Mutation: {},
    Subscription: {},
    typeDefs: [],
  },
)

const { name, typeDefs, ...resolverData } = resolvers
log.create('query')(resolverData.Query)
log.create('resolver')(resolverData.Resolver)
log.create('mutation')(resolverData.Mutation)
log.create('subscription')(resolverData.Subscription)
if (!1) log(typeDefs.join('\n'))
