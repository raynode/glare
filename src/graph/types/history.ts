import { Build } from 'gram'
import { getMemory } from 'services/memory'

const stack = []
const stackSize = 100
const stackUpdaterInterval = 5000 // 5s
let interval: ReturnType<typeof setInterval> | false = false

const updater = () => {
  stack.unshift({
    time: Date.now(),
    ...getMemory(),
  })
  if (stack.length - 1 > stackSize) stack.pop()
}

export const startUpdater = () => {
  if (interval)
    // interval running
    return
  interval = setInterval(updater, stackUpdaterInterval)
}

export const stopUpdater = () => {
  if (interval) {
    clearInterval(interval)
    interval = false
  }
}

export const historyBuild = <BuildMode, Context>(build: Build<BuildMode, Context>) => {
  build.addQuery('history', 'History!', { resolver: () => ({}) })
  build.addType('HistoryMemory', {
    type: 'type',
    fields: {
      time: 'DateTime!',
      rss: 'Int!',
      heapTotal: 'Int!',
      heapUsed: 'Int!',
      external: 'Int!',
    },
  })
  build.addType('History', {
    fields: {
      memory: {
        args: {
          limit: 'Int',
        },
        type: '[HistoryMemory!]!',
      },
    },
    resolver: {
      memory: (_, args, context) => {
        const limit = (args && args.limit) || stackSize
        return stack.slice(0, limit)
      },
    },
  })
}
