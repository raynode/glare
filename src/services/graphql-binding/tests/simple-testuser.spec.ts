// tslint:disable-next-line
import * as faker from 'faker'

import { createModel } from '../create-model'

import { Methods, QueryGenerator } from '../types'
import { modelsToSchema, modelsToSDL } from './test-utils'

import { graphql, GraphQLSchema } from 'graphql'

interface TestUser {
  id: number
  email: string
  nickname?: string
}
let testusers = []
const queryGenerator: QueryGenerator = <T>(model: T) => {
  const methods: Methods<TestUser> = {
    createOne: async user => {
      testusers.push(user)
      return user as TestUser
    },
    findOne: async ({ nickname, email }, order, offset) => {
      if (nickname && email) return testusers.find(user => user.nickname === nickname && user.email === email)
      if (nickname) return testusers.find(user => user.nickname === nickname)
      if (email) return testusers.find(user => user.email === email)
      throw new Error('need some filter')
    },
    deleteMany: async (where, order, offset, limit) => {
      const hits = await methods.findMany(where, order, offset, limit)
      const ids = hits.map(user => user.id)
      testusers = testusers.filter(user => !ids.includes(user.id))
      return hits
    },
    findMany: async ({ nickname, email }, order, offset, limit) => {
      if (nickname && email) return testusers.filter(user => user.nickname === nickname && user.email === email)
      if (nickname) return testusers.filter(user => user.nickname === nickname)
      if (email) return testusers.filter(user => user.email === email)
      throw new Error('need some filter')
    },
    update: async (where, order, offset, { nickname, email }) => {
      const user = await methods.findOne(where, order, offset, 1)
      if (user) {
        if (nickname) user.nickname = nickname
        if (email) user.email = email
      }
      return user
    },
  }
  return methods as any
}

const Testuser = createModel('Testuser', {
  id: {
    allowNull: false,
    allowUpdate: true,
    comment: 'Invisible Field, hidden from the schema',
    type: 'string',
    visible: false,
  },
  email: {
    allowNull: false,
    allowUpdate: true,
    comment: 'Not-Null field',
    type: 'string',
    visible: true,
  },
  nickname: {
    allowNull: true,
    allowUpdate: true,
    comment: 'Nullable field',
    defaultValue: null,
    type: 'string',
    visible: true,
  },
})

describe('basic graphql generation', () => {
  let run: (query: string, variables?: Record<string, any>, context?: any) => any
  let schema: GraphQLSchema

  beforeEach(() => {
    testusers = [
      {
        id: 473,
        email: 'test@domain.com',
        nickname: 'example',
      },
    ]
    schema = modelsToSchema([Testuser], { methodMapper: queryGenerator })
    run = (query, variables = {}, ctx = null) => graphql(schema, query, null, { ctx }, variables)
  })

  it('should create a basic graphql setup from a model', () => {
    expect(modelsToSDL([Testuser])).toMatchSnapshot()
  })

  it('should run a findOne query correctly', async () => {
    const result = await run(`{
      data: Testuser(where: { nickname: "example" }) { email }
    }`)
    expect(result.data).toEqual({ data: { email: 'test@domain.com' } })
  })

  it('should run a findMany query correctly', async () => {
    const result = await run(`{
      data: Testusers(where: { email: "test@domain.com" }) { nickname }
    }`)
    expect(result.data).toEqual({ data: [{ nickname: 'example' }] })
  })

  it('should run a create mutation correctly', async () => {
    const result = await run(`mutation {
      data: createTestuser(data: { email: "new+user@domain.com", nickname: "newb" }) { email nickname }
    }`)
    expect(result.data).toEqual({ data: { email: 'new+user@domain.com', nickname: 'newb' } })
    expect(testusers).toHaveLength(2)
  })

  it('should run a delete mutation correctly', async () => {
    const result = await run(`mutation {
      data: deleteTestusers(where: { nickname: "example" }) { email nickname }
    }`)
    expect(result.data).toEqual({ data: [{ email: 'test@domain.com', nickname: 'example' }] })
    expect(testusers).toHaveLength(0)
  })

  it('should run a update mutation correctly', async () => {
    const result = await run(`mutation {
      data: updateTestuser(
        where: { nickname: "example" }
        data: {
          email: "super+user@domain.com"
          nickname: "superman"
        }
      ) { email nickname }
    }`)
    expect(result.data).toEqual({ data: { email: 'super+user@domain.com', nickname: 'superman' } })
  })
})
