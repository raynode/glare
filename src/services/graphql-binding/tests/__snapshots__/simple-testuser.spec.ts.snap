// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic graphql generation should create a basic graphql setup from a model 1`] = `
"type Mutation {
  createTestuser(data: TestuserCreateInput!): Testuser
  updateTestuser(where: TestuserFilter!, data: TestuserUpdateInput!): Testuser
  deleteTestusers(where: TestuserFilter!): [Testuser!]!
}

input Page {
  limit: Int
  offset: Int
}

type Query {
  Testuser(where: TestuserFilter!, order: TestuserOrderEnum): Testuser
  Testusers(where: TestuserFilter, order: TestuserOrderEnum, page: Page): [Testuser!]!
}

type Testuser {
  email: String!
  nickname: String
}

input TestuserCreateInput {
  email: String!
  nickname: String
}

input TestuserFilter {
  email: String
  nickname: String
}

enum TestuserOrderEnum {
  email_ASC
  email_DESC
  nickname_ASC
  nickname_DESC
}

input TestuserUpdateInput {
  email: String
  nickname: String
}
"
`;
